File: backend-api/proto/draft.py
# draft.py
import sys, readline
import pandas as pd
from tabulate import tabulate
from util import load_players, positional_lists, load_espn_ranks, attach_espn_ranks_inplace, report_espn_match_coverage, load_adp, attach_adp_inplace
from models import DraftState, LINEUP
import numpy as np

from demand import (
    forecast_until_next_pick_esbn,
    survival_probs,
    next_pick_probs_mix  # add to your imports
)
from var import expected_best_next, current_best_now, league_replacement_indices, replacement_ppg_by_pos, best_ix_by_pos_now, davar_esbn_deficit_weighted
from util import positional_lists

from collections import defaultdict # ensure this import exists

def _espn_rank_col(df):
    # prefer ESPN rank; fall back to any global rank you keep
    return 'espn_rank' if 'espn_rank' in df.columns else 'global_rank'

def current_espn_board_positions(df, available_ix):
    """
    Return {player_ix: 1-based ordinal} on the CURRENT ESPN board,
    i.e., sort remaining players by ESPN rank and number them 1..K.
    Example: if a player was #5 pre-draft and top 4 are gone -> they are #1.
    """
    col = _espn_rank_col(df)
    avail = df.loc[list(available_ix)].sort_values(col)
    return {ix: i+1 for i, ix in enumerate(avail.index)}


def available_indices(df, taken):
    return [i for i in df.index if i not in taken]

def predict_current_pick(df, draft):
    """
    Predict the player the CURRENT team (draft.current_pick owner) will take,
    using the ESPN hazard for THIS pick only.
    Returns: (pred_ix, hazard_current_dict)
    """
    owner = draft.pick_owner(draft.current_pick)
    team = draft.teams[owner]
    avail_ix = available_indices(df, draft.taken)

    probs = next_pick_probs_mix(
        df, avail_ix, draft.current_pick, team
    )

    picked_player = np.random.choice(
        list(probs.keys()), 
        p=list(probs.values())
    )

    return picked_player, probs

def _assign_user_lineup(df, team):
    """
    Greedy, best-ball style assignment for DISPLAY ONLY:
    - Fill starters at QB/RB/WR/TE by highest PPG.
    - Fill FLEX from remaining RB/WR/TE by highest PPG.
    - Remaining players shown as BENCH (sorted by PPG).
    """
    by_pos = defaultdict(list)
    for ix in team.picks:
        pos = df.loc[ix, 'position']
        by_pos[pos].append(ix)
    for pos in by_pos:
        by_pos[pos].sort(key=lambda i: float(df.loc[i, 'ppg']), reverse=True)

    rows, assigned = [], set()

    # Starters
    for pos in ['QB', 'RB', 'WR', 'TE']:
        filled = LINEUP.get(pos, 0) - team.need.get(pos, 0)  # slots actually filled so far
        # take top 'filled' players at this pos as starters
        for i, ix in enumerate(by_pos.get(pos, [])[:max(filled, 0)]):
            assigned.add(ix)
            rows.append([f"{pos}{i+1}", df.loc[ix, 'player_name'], pos, round(float(df.loc[ix, 'ppg']), 2)])

    # FLEX (from extras across RB/WR/TE)
    flex_filled = LINEUP.get('FLEX', 0) - team.need.get('FLEX', 0)
    flex_pool = []
    for pos in ['RB', 'WR', 'TE']:
        starters = LINEUP.get(pos, 0)
        extras = by_pos.get(pos, [])[starters:]  # players beyond starter count
        flex_pool.extend(extras)
    flex_pool.sort(key=lambda i: float(df.loc[i, 'ppg']), reverse=True)
    for i, ix in enumerate(flex_pool[:max(flex_filled, 0)]):
        assigned.add(ix)
        rows.append([f"FLEX{i+1}", df.loc[ix, 'player_name'], df.loc[ix, 'position'], round(float(df.loc[ix, 'ppg']), 2)])

    # Bench (anything not assigned above)
    bench = []
    for pos, lst in by_pos.items():
        for ix in lst:
            if ix not in assigned:
                bench.append(ix)
    bench.sort(key=lambda i: float(df.loc[i, 'ppg']), reverse=True)
    for ix in bench:
        rows.append(["BENCH", df.loc[ix, 'player_name'], df.loc[ix, 'position'], round(float(df.loc[ix, 'ppg']), 2)])

    return rows

def print_user_roster(df, draft, team_names=None):
    team = draft.teams[draft.user_team_ix]
    rows = _assign_user_lineup(df, team)
    counts = []
    for pos in ['QB', 'RB', 'WR', 'TE', 'FLEX']:
        filled = LINEUP.get(pos, 0) - team.need.get(pos, 0)
        remaining = team.need.get(pos, 0)
        counts.append([pos, max(filled, 0), max(remaining, 0)])

    team_name = team_names[draft.user_team_ix] if team_names else f"Team {draft.user_team_ix}"
    print(f"\n-- {team_name}'s roster so far --")
    if rows:
        print(tabulate(rows, headers=["Slot", "Player", "Pos", "PPG"], tablefmt="github"))
    else:
        print("(no picks yet)")
    print(tabulate(counts, headers=["Slot", "Filled", "Remaining"], tablefmt="github"))


def available_indices(df, taken):
    return [i for i in df.index if i not in taken]

def steps_until_user_next_pick(draft):
    cur = draft.current_pick
    total = draft.n_teams * draft.rounds
    steps = 0
    # if it's your pick right now, look AFTER you pick
    if draft.pick_owner(cur) == draft.user_team_ix:
        cur += 1  # skip your current pick (e.g., at 24, next is 25)
    # include the current non-user pick in the count
    while (cur + steps) <= total and draft.pick_owner(cur + steps) != draft.user_team_ix:
        steps += 1
    return steps

def adp_value_tag(nfc_adp, ref_pick_num):
    """Return a short tag like 'value +10', 'reach 15', or 'at ADP'."""
    try:
        if pd.isna(nfc_adp) or nfc_adp is None:
            return "N/A"
        adp = float(nfc_adp)
    except (TypeError, ValueError):
        return "N/A"  # unknown

    if not pd.notna(adp):
        return "N/A"

    diff = int(round(ref_pick_num - adp))  # >0 => value; <0 => reach
    if diff > 0:
        return f"value +{diff}"
    elif diff < 0:
        return f"reach {abs(diff)}"
    else:
        return "at ADP"


def show_recs(df, draft, topN=12, alpha=0.9, beta=0.6):
    # horizon until user's next pick
    h = steps_until_user_next_pick(draft)
    avail_ix = available_indices(df, draft.taken)
    pos_lists = positional_lists(df)

    team = draft.teams[draft.user_team_ix]
    ref_pick_num = draft.current_pick

    cap_left_by_pos = {}
    for pos in ['QB', 'RB', 'WR', 'TE']:
        cap = float(team.need.get(pos, 0))
        if pos in ('RB', 'WR', 'TE'):
            cap += float(team.need.get('FLEX', 0)) / 3.0

        cap_left_by_pos[pos] = cap
        
    owned_count_by_pos = {pos: 0 for pos in ['QB', 'RB', 'WR', 'TE']}
    for ix in team.picks:
        owned_count_by_pos[df.loc[ix, 'position']] += 1


    # ESPN-based hazards & drains
    hazards, E_drain = forecast_until_next_pick_esbn(df, draft, avail_ix, h)

    print(f"Horizon to your next pick (H) = {h}, sum(E[K_pos]) = {round(sum(E_drain.values()),2)}")
    
    board_pos_map = current_espn_board_positions(df, avail_ix)
    
    # Survival for candidate set (take top ~60 by PPG among available)
    cand_df = df.loc[avail_ix].sort_values('ppg', ascending=False).head(60)
    surv = survival_probs(hazards, cand_df.index)

    best_ix_map = best_ix_by_pos_now(pos_lists, avail_ix)
    surv_best_by_pos = {p: (surv.get(ix, 1.0) if ix is not None else 1.0) 
                        for p, ix in best_ix_map.items()}


    # Best-now and expected-best-next by position
    best_now = current_best_now(df, pos_lists, avail_ix)
    E_best_next = expected_best_next(df, pos_lists, avail_ix, E_drain)

    # Replacement indices
    repl_idx_map = league_replacement_indices(draft.n_teams, LINEUP)
    repl_ppg = replacement_ppg_by_pos(df, pos_lists, avail_ix, repl_idx_map)


    from var import remaining_var_budget_by_pos, accrued_var_by_pos

    # how many bench slots remain in THIS league/team
    bench_left_now = draft.teams[draft.user_team_ix].bench_left()

    # accrued VAR so far (starters + already-filled FLEX)
    A_by_pos = accrued_var_by_pos(df, draft.teams[draft.user_team_ix], LINEUP, repl_ppg)

    # remaining budget with bench awareness
    R_by_pos = remaining_var_budget_by_pos(
        df, pos_lists, avail_ix, draft.teams[draft.user_team_ix], LINEUP, repl_ppg,
        bench_left=bench_left_now, bench_weight=1.0  # tune weight if desired
)

    # Build table
    rows = []
    for ix, r in cand_df.iterrows():

        score = davar_esbn_deficit_weighted(
            df, ix,
            best_now, E_best_next, repl_ppg, surv_best_by_pos,
            survival_for_candidate=surv.get(ix, 1.0),
            accrued_var_by_pos=A_by_pos,
            remain_var_budget_by_pos=R_by_pos,
            alpha=0.9, beta=0.7,
            cap_left_by_pos=cap_left_by_pos,
            owned_count_by_pos=owned_count_by_pos,
            rho_overfill=1.0,
            phi_deficit=1.0
        )

        adp_tag = adp_value_tag(r.get('nfc_adp', None), ref_pick_num)

        rows.append([
            r['player_name'],
            r['position'],
            board_pos_map.get(ix, None),                       # <-- NEW: current ESPN board #
            round(float(r['ppg']), 2),
            f"{100*surv.get(ix,1.0):.0f}%",
            adp_tag,
            round(E_best_next[r['position']] - best_now[r['position']], 2),
            round(score, 2),
            ix
        ])

    rows.sort(key=lambda x: x[7], reverse=True)

    from tabulate import tabulate
    print("\n== Recommendations (as if it's YOUR pick next) ==")
    print(tabulate(
    rows[:topN],
    headers=["Player","Pos","ESPN#","PPG","Survive%","ADP","OppCost(Pos)","DAVAR","idx"],
    tablefmt="github"
    ))

    print("\nPos drain by your next pick (E[K_pos]):", {k: round(v,2) for k,v in E_drain.items()})

    owner_now = draft.pick_owner(draft.current_pick)
    pred_ix, hazard_current = predict_current_pick(df, draft)
    if pred_ix is not None:
        p = df.loc[pred_ix]
        prob = hazard_current.get(pred_ix, 0.0)
        print(f"\nModel predicts Team {owner_now} will pick: "
              f"{p['player_name']} ({p['position']}, {float(p['ppg']):.2f} PPG) — p={prob:.1%}")

    # --- NEW: your top recommendation (from the table you just saw) ---
    if rows:
        best_row = rows[0]
        print(f"Top rec for YOUR next turn: {best_row[0]} ({best_row[1]}) — "
              f"DAVAR {best_row[7]:.2f}, Survive {best_row[4]}")   

    return rows, hazards, E_drain, pred_ix

DATA = "data/player_rankings.csv"
ESPN = "data/espn_rankings_final.csv"
ADP = "data/ppr_adp_new.csv"
ROUNDS = 15  # default rounds, can be made configurable later

def available_indices(df, taken):
    return [i for i in df.index if i not in taken]

def find_player(df, query):
    cand = df[df['player_name'].str.lower().str.contains(query.lower())]
    return list(cand.index)

def get_draft_config():
    """Get draft configuration from user"""
    print("=== Fantasy Football Draft Setup ===")
    
    # Get number of teams
    while True:
        try:
            n_teams = int(input("Enter number of teams in the draft: "))
            if n_teams >= 2 and n_teams <= 16:
                break
            else:
                print("Please enter a number between 2 and 16.")
        except ValueError:
            print("Please enter a valid number.")
    
    # Get number of rounds
    while True:
        try:
            rounds = int(input(f"Enter number of rounds (default {ROUNDS}): ") or ROUNDS)
            if rounds >= 1 and rounds <= 20:
                break
            else:
                print("Please enter a number between 1 and 20.")
        except ValueError:
            print("Please enter a valid number.")
    
    return n_teams, rounds

def get_team_names(n_teams):
    """Get team names from user in draft order (1st pick to last pick)"""
    print(f"\n=== Enter Team Names in Draft Order ===")
    print(f"Enter the names of {n_teams} teams in the order they will draft (1st pick to last pick)")
    print("This will be mapped to the snake draft order automatically.\n")
    
    team_names = []
    for i in range(n_teams):
        while True:
            name = input(f"Team {i+1} (pick #{i+1}): ").strip()
            if name:
                team_names.append(name)
                break
            print("Please enter a valid team name.")
    
    return team_names

def get_snake_draft_order(n_teams, rounds):
    """Generate the complete snake draft order"""
    draft_order = []
    for round_num in range(1, rounds + 1):
        if round_num % 2 == 1:  # Odd rounds: 1, 2, 3, ..., n
            round_order = list(range(n_teams))
        else:  # Even rounds: n, n-1, ..., 2, 1
            round_order = list(range(n_teams - 1, -1, -1))
        
        for team_idx in round_order:
            draft_order.append(team_idx)
    
    return draft_order

def print_draft_order(team_names, draft_order):
    """Print the complete draft order for reference"""
    print(f"\n=== Complete Draft Order ===")
    for i, team_idx in enumerate(draft_order):
        round_num = (i // len(team_names)) + 1
        pick_in_round = (i % len(team_names)) + 1
        print(f"Pick {i+1:2d} (Round {round_num:2d}.{pick_in_round:2d}): {team_names[team_idx]}")
    print()

def main():
    df = load_players(DATA)

    # Attach ESPN ranks (creates df['espn_rank'])
    try:
        espn = load_espn_ranks(ESPN)
        attach_espn_ranks_inplace(df, espn)
        report_espn_match_coverage(df)

    except Exception as e:
        print(f"[ESPN merge] Warning: {e} — falling back to global_rank for hazards.")

    adp = load_adp(ADP)
    attach_adp_inplace(df, adp)

    # Get draft configuration from user
    n_teams, rounds = get_draft_config()
    
    # Get team names from user
    team_names = get_team_names(n_teams)
    
    # Generate snake draft order
    draft_order = get_snake_draft_order(n_teams, rounds)
    
    # Print the complete draft order for reference
    print_draft_order(team_names, draft_order)
    
    # Ask user which team they are
    print("Which team are you?")
    for i, name in enumerate(team_names):
        print(f"{i+1}. {name}")
    
    while True:
        try:
            user_choice = int(input(f"Enter your team number (1-{n_teams}): ")) - 1
            if 0 <= user_choice < n_teams:
                user_team = user_choice
                break
            else:
                print(f"Please enter a number between 1 and {n_teams}")
        except ValueError:
            print("Please enter a valid number")

    draft = DraftState(n_teams=n_teams, rounds=rounds, user_team_ix=user_team)

    while not draft.is_complete():
        owner = draft.pick_owner(draft.current_pick)
        owner_name = team_names[owner]

        print_user_roster(df, draft, team_names)

        # Always show recs as if it's YOUR pick next (returns hazards for upcoming h picks)
        rows, hazards, E_drain, pred_ix = show_recs(
            df, draft,
            topN=24,         # how many rows to display
            alpha=0.9,       # DAVAR weight on pos-wait cost
            beta=0.6         # DAVAR cross-pos hedge weight
        )

        print(f"\nPick {draft.current_pick} is {owner_name} (Team {owner}).")
        cmd = input("Enter pick: 'name' to pick by search, 'idx' to pick by idx, or 'auto' to auto-pick for owner: ").strip()

        if cmd.lower() == 'auto':
            if pred_ix is None:
                pred_ix, _ = predict_current_pick(df, draft)
            pick_ix = pred_ix
            
        elif cmd.isdigit():
            pick_ix = int(cmd)

        else:
            matches = find_player(df, cmd)
            if not matches:
                print("No matches; try again."); continue
            # choose top by PPG among matches
            pick_ix = df.loc[matches].sort_values('ppg', ascending=False).index[0]

        # Validate roster need
        pos = df.loc[pick_ix, 'position']
        if not draft.teams[owner].can_draft(pos):
            print(f"{owner_name} cannot draft {pos} (slots full). Try another pick.")
            continue

        # Apply pick
        draft.taken.add(pick_ix)
        draft.teams[owner].picks.append(pick_ix)
        draft.teams[owner].add_player(pos)
        # NOTE: No Dirichlet/Bayesian update needed in ESPN hazard model.

        # Advance
        draft.current_pick += 1

    print("\nDraft complete!")
    u = draft.teams[draft.user_team_ix]
    roster = [[df.loc[ix,'player_name'], df.loc[ix,'position'], round(df.loc[ix,'ppg'],2)] for ix in u.picks]
    print(f"\n=== Final Roster for {team_names[user_team]} ===")
    print(tabulate(roster, headers=["Player","Pos","PPG"], tablefmt="github"))


if __name__ == "__main__":
    main()

File: backend-api/src/algorithm/draft.ts
/**
 * Main draft recommendation engine
 * TypeScript port of draft.py show_recs() function and related functionality
 */

import { 
  Player, 
  Position, 
  DraftState, 
  DraftRecommendation, 
  createPositionLists,
  isValidPosition 
} from './models';

import {
  getCurrentBestNow,
  getExpectedBestNext,
  getLeagueReplacementIndices,
  getReplacementPpgByPosition,
  getCurrentEspnBoardPositions,
  getAdpValueTag,
  roundTo
} from './utils';

import {
  forecastUntilNextPickEsbn,
  getSurvivalProbabilities,
  predictCurrentPick,
  getBestIndexByPositionNow
} from './survival';

import {
  calculateDavarEsbn,
  scorePlayersWithDavar,
  calculateAllOpportunityCosts,
  ScoringContext
} from './scoring';

/**
 * Main recommendation engine
 * Port of show_recs() from draft.py
 */
export interface RecommendationOptions {
  topN?: number;        // Number of recommendations to return
  alpha?: number;       // DAVAR weight on position-wait cost  
  beta?: number;        // DAVAR cross-position hedge weight
  candidatePoolSize?: number; // Size of candidate pool to consider
}

export interface RecommendationResult {
  recommendations: DraftRecommendation[];
  horizon: number;
  positionDrain: Record<Position, number>;
  predictedPick?: {
    playerIndex: number;
    playerName: string;
    position: Position;
    probability: number;
  };
  debug?: {
    availableCount: number;
    candidatePoolSize: number;
    bestNowByPosition: Record<Position, number>;
    expectedBestNextByPosition: Record<Position, number>;
    replacementPpgByPosition: Record<Position, number>;
  };
}

export function generateRecommendations(
  players: Map<number, Player>,
  draftState: DraftState,
  options: RecommendationOptions = {}
): RecommendationResult {
  const {
    topN = 12,
    alpha = 0.9,
    beta = 0.6,
    candidatePoolSize = 60
  } = options;

  // Calculate horizon until user's next pick
  const horizon = draftState.stepsUntilUserNextPick();
  
  // Get available player indices
  const allIndices = Array.from(players.keys());
  const availableIndices = draftState.availableIndices(allIndices);
  
  // Create position lists from available players
  const availablePlayers = availableIndices
    .map(idx => players.get(idx)!)
    .filter(p => p !== undefined);
  
  const positionLists = createPositionLists(availablePlayers);

  // ESPN-based hazards & position drains
  const { hazards, expectedDrain } = forecastUntilNextPickEsbn(
    players,
    draftState,
    availableIndices,
    horizon
  );

  // Current ESPN board positions
  const boardPositionMap = getCurrentEspnBoardPositions(players, availableIndices);
  
  // Select candidate pool (top players by PPG among available)
  const candidatePool = availablePlayers
    .sort((a, b) => b.ppg - a.ppg)
    .slice(0, candidatePoolSize);

  // Calculate survival probabilities for candidates
  const candidateIndices = candidatePool.map(p => p.index);
  const survivalProbs = getSurvivalProbabilities(hazards, candidateIndices);

  // Calculate current best and expected best next by position
  const bestNowByPosition = getCurrentBestNow(players, positionLists, availableIndices);
  const expectedBestNextByPosition = getExpectedBestNext(
    players,
    positionLists,
    availableIndices,
    expectedDrain
  );

  // Calculate replacement level players
  const replacementIndices = getLeagueReplacementIndices(
    draftState.nTeams,
    { QB: 1, RB: 2, WR: 2, TE: 1, FLEX: 1 }
  );

  const replacementPpgByPosition = getReplacementPpgByPosition(
    players,
    positionLists,
    availableIndices,
    replacementIndices
  );

  // Score all candidates using DAVAR
  const scoringContext: ScoringContext = {
    bestNowByPosition,
    expectedBestNextByPosition,
    replacementPpgByPosition,
    survivalProbabilities: survivalProbs,
    options: { alpha, beta }
  };

  const scoredPlayers = scorePlayersWithDavar(candidatePool, scoringContext);

  // Calculate opportunity costs by position
  const opportunityCosts = calculateAllOpportunityCosts(
    bestNowByPosition,
    expectedBestNextByPosition
  );

  // Build recommendation list
  const recommendations: DraftRecommendation[] = scoredPlayers
    .slice(0, topN)
    .map(result => {
      const player = result.player;
      const espnBoardPos = boardPositionMap[player.index] || null;
      const adpValue = getAdpValueTag(player.adp, draftState.currentPick);
      const opportunityCost = opportunityCosts[player.position];

      return {
        playerIndex: player.index,
        playerName: player.playerName,
        position: player.position,
        espnBoardPosition: espnBoardPos,
        ppg: roundTo(player.ppg, 2),
        survivalPercent: `${Math.round(result.survivalProbability * 100)}%`,
        adpValue: adpValue,
        opportunityCost: roundTo(opportunityCost, 2),
        davarScore: roundTo(result.davarScore, 2)
      };
    });

  // Predict current pick
  let predictedPick: RecommendationResult['predictedPick'] = undefined;
  const prediction = predictCurrentPick(players, draftState, availableIndices);
  
  if (prediction.predictedIndex !== null) {
    const predictedPlayer = players.get(prediction.predictedIndex);
    if (predictedPlayer) {
      const maxProb = Math.max(...Object.values(prediction.probabilities));
      predictedPick = {
        playerIndex: prediction.predictedIndex,
        playerName: predictedPlayer.playerName,
        position: predictedPlayer.position,
        probability: maxProb
      };
    }
  }

  return {
    recommendations,
    horizon,
    positionDrain: expectedDrain,
    predictedPick,
    debug: {
      availableCount: availableIndices.length,
      candidatePoolSize: candidatePool.length,
      bestNowByPosition,
      expectedBestNextByPosition,
      replacementPpgByPosition
    }
  };
}

/**
 * Find players by search query
 * Port of find_player() from draft.py
 */
export function findPlayers(
  players: Map<number, Player>,
  query: string
): Player[] {
  if (!query || query.trim().length === 0) {
    return [];
  }

  const lowerQuery = query.toLowerCase().trim();
  const results: Player[] = [];

  for (const player of players.values()) {
    if (player.playerName.toLowerCase().includes(lowerQuery)) {
      results.push(player);
    }
  }

  // Sort by PPG descending
  return results.sort((a, b) => b.ppg - a.ppg);
}

/**
 * Get available player indices (not yet drafted)
 * Port of available_indices() from draft.py
 */
export function getAvailableIndices(
  players: Map<number, Player>,
  takenIndices: Set<number>
): number[] {
  return Array.from(players.keys()).filter(idx => !takenIndices.has(idx));
}

/**
 * Validate if a pick is legal for a team
 */
export function validatePick(
  playerIndex: number,
  players: Map<number, Player>,
  draftState: DraftState,
  teamIndex: number
): {
  valid: boolean;
  reason?: string;
} {
  const player = players.get(playerIndex);
  if (!player) {
    return { valid: false, reason: "Player not found" };
  }

  if (draftState.taken.has(playerIndex)) {
    return { valid: false, reason: "Player already drafted" };
  }

  if (!isValidPosition(player.position)) {
    return { valid: false, reason: "Invalid position" };
  }

  const team = draftState.teams[teamIndex];
  if (!team.canDraft(player.position)) {
    return { 
      valid: false, 
      reason: `Team cannot draft ${player.position} (roster slots full)` 
    };
  }

  return { valid: true };
}

/**
 * Apply a draft pick to the draft state
 */
export function applyDraftPick(
  playerIndex: number,
  players: Map<number, Player>,
  draftState: DraftState
): {
  success: boolean;
  message?: string;
  pickDetails?: {
    playerName: string;
    position: Position;
    team: number;
    pickNumber: number;
  };
} {
  const player = players.get(playerIndex);
  if (!player) {
    return { success: false, message: "Player not found" };
  }

  // Determine which team owns this pick
  const teamIndex = draftState.pickOwner(draftState.currentPick);

  // Validate the pick
  const validation = validatePick(playerIndex, players, draftState, teamIndex);
  if (!validation.valid) {
    return { success: false, message: validation.reason };
  }

  // Apply the pick
  draftState.taken.add(playerIndex);
  draftState.teams[teamIndex].picks.push(playerIndex);
  draftState.teams[teamIndex].addPlayer(player.position);
  
  const pickNumber = draftState.currentPick;
  draftState.currentPick++;

  return {
    success: true,
    pickDetails: {
      playerName: player.playerName,
      position: player.position,
      team: teamIndex,
      pickNumber
    }
  };
}

/**
 * Generate a simple text summary of recommendations
 * Similar to the console output from the Python version
 */
export function formatRecommendationSummary(
  result: RecommendationResult,
  draftState: DraftState
): string {
  const lines: string[] = [];
  
  lines.push(`Horizon to your next pick (H) = ${result.horizon}`);
  lines.push(`Sum of E[K_pos] = ${roundTo(Object.values(result.positionDrain).reduce((a, b) => a + b, 0), 2)}`);
  lines.push('');
  lines.push('== Recommendations (as if it\'s YOUR pick next) ==');
  
  // Header
  lines.push('Player\t\tPos\tESPN#\tPPG\tSurvive%\tADP\t\tOppCost\tDAVAR');
  lines.push('-'.repeat(80));
  
  // Recommendations
  for (const rec of result.recommendations) {
    const espn = rec.espnBoardPosition?.toString() || '-';
    const line = [
      rec.playerName.padEnd(16),
      rec.position,
      espn.padStart(5),
      rec.ppg.toString().padStart(5),
      rec.survivalPercent.padStart(8),
      rec.adpValue.padEnd(10),
      rec.opportunityCost.toString().padStart(7),
      rec.davarScore.toString().padStart(6)
    ].join('\t');
    lines.push(line);
  }
  
  lines.push('');
  lines.push('Position drain by your next pick:');
  for (const [pos, drain] of Object.entries(result.positionDrain)) {
    lines.push(`  ${pos}: ${roundTo(drain, 2)}`);
  }
  
  if (result.predictedPick) {
    const owner = draftState.pickOwner(draftState.currentPick);
    lines.push('');
    lines.push(`Model predicts Team ${owner} will pick: ${result.predictedPick.playerName} ` +
               `(${result.predictedPick.position}) — p=${Math.round(result.predictedPick.probability * 100)}%`);
  }
  
  if (result.recommendations.length > 0) {
    const topRec = result.recommendations[0];
    lines.push('');
    lines.push(`Top rec for YOUR next turn: ${topRec.playerName} (${topRec.position}) — ` +
               `DAVAR ${topRec.davarScore}, Survive ${topRec.survivalPercent}`);
  }
  
  return lines.join('\n');
}

/**
 * Create a sample draft state for testing
 */
export function createSampleDraftState(
  nTeams: number = 12,
  rounds: number = 15,
  userTeamIndex: number = 0,
  currentPick: number = 1
): DraftState {
  const draftState = new DraftState(nTeams, rounds, userTeamIndex);
  draftState.currentPick = currentPick;
  return draftState;
}

/**
 * Load players from a data structure
 * Helper for creating the player map from various data sources
 */
export function createPlayersMap(playerData: Array<{
  index: number;
  playerName: string;
  position: string;
  team?: string | null;
  ppg: number;
  adp?: number;
  globalRank?: number;
  espnRank?: number;
}>): Map<number, Player> {
  const players = new Map<number, Player>();

  for (const data of playerData) {
    if (!isValidPosition(data.position)) {
      continue; // Skip invalid positions
    }

    const player: Player = {
      index: data.index,
      playerName: data.playerName,
      position: data.position as Position,
      team: data.team || null,
      ppg: data.ppg,
      adp: data.adp,
      globalRank: data.globalRank,
      espnRank: data.espnRank
    };

    players.set(data.index, player);
  }

  return players;
}
